File: bindings/python/src/graphqlite/algorithms/_parsing.py
"""Shared parsing helpers for algorithm results."""

from typing import Any, List, Optional


# Known column names for graph algorithm results
ALGO_COLUMN_NAMES = [
    "column_0", "wcc()", "scc()", "pagerank()", "degree_centrality()",
    "betweenness_centrality()", "closeness_centrality()", "eigenvector_centrality()",
    "labelPropagation()", "louvain()", "bfs()", "dfs()", "apsp()"
]


def extract_algo_array(result: List[dict]) -> List[dict]:
    """Extract wrapped array results from graph algorithms.

    Graph algorithms return results in one of two formats:
    1. Old format: Multiple rows with fields directly accessible
    2. New format: Single row with a column containing an array of objects

    This function detects the new format and extracts the array elements.
    """
    # If multiple rows, assume old format - return as-is
    if len(result) != 1:
        return result

    # Single row - check if it has an array column
    row = result[0]

    # Try common column names for wrapped array results
    for col_name in ALGO_COLUMN_NAMES:
        if col_name in row and isinstance(row[col_name], list):
            return row[col_name]

    # No array column found, return original result
    return result


def parse_score_result(row: dict, score_key: str = "score") -> Optional[dict]:
    """Parse a result row with node_id, user_id, and a score field."""
    node_id = row.get("node_id")
    user_id = row.get("user_id")
    score = row.get(score_key)

    if node_id is None:
        return None

    return {
        "node_id": str(node_id),
        "user_id": user_id,
        "score": float(score) if score is not None else 0.0
    }


def parse_community_result(row: dict) -> Optional[dict]:
    """Parse a community detection result row."""
    node_id = row.get("node_id")
    user_id = row.get("user_id")
    community = row.get("community")

    if node_id is None or community is None:
        return None

    return {
        "node_id": str(node_id),
        "user_id": user_id,
        "community": int(community) if community else 0
    }


def parse_component_result(row: dict) -> Optional[dict]:
    """Parse a connected components result row."""
    node_id = row.get("node_id")
    user_id = row.get("user_id")
    component = row.get("component")

    if node_id is None:
        return None

    return {
        "node_id": str(node_id),
        "user_id": user_id,
        "component": int(component) if component is not None else 0
    }


def parse_traversal_result(row: dict) -> Optional[dict]:
    """Parse a BFS/DFS traversal result row."""
    user_id = row.get("user_id")
    depth = row.get("depth")
    order = row.get("order")

    if user_id is None:
        return None

    return {
        "user_id": user_id,
        "depth": int(depth) if depth is not None else 0,
        "order": int(order) if order is not None else 0
    }


def safe_float(val: Any, default: float = 0.0) -> float:
    """Safely convert a value to float."""
    if val is None:
        return default
    try:
        return float(val)
    except (ValueError, TypeError):
        return default


def safe_int(val: Any, default: int = 0) -> int:
    """Safely convert a value to int."""
    if val is None:
        return default
    try:
        return int(val)
    except (ValueError, TypeError):
        return default

File: bindings/rust/src/algorithms/parsing.rs
//! Shared parsing helpers for algorithm results.

use crate::{Row, Value};

/// Extract a wrapped array result from graph algorithms.
///
/// Graph algorithms return results in one of two formats:
/// 1. Old format: Multiple rows with fields directly accessible
/// 2. New format: Single row with a column containing an array of objects
///
/// This function detects the new format and extracts the array elements,
/// converting each object into a Row for consistent processing.
pub(crate) fn extract_algo_array(result: &[&Row]) -> Vec<Row> {
    // If multiple rows, assume old format - clone and return
    if result.len() != 1 {
        return result.iter().map(|r| (*r).clone()).collect();
    }

    // Single row - check if it has an array column
    let row = result[0];

    // Try common column names for wrapped array results
    for col_name in [
        "column_0",
        "wcc()",
        "scc()",
        "pagerank()",
        "degree_centrality()",
        "betweenness_centrality()",
        "closeness_centrality()",
        "eigenvector_centrality()",
        "labelPropagation()",
        "louvain()",
    ] {
        if let Some(Value::Array(arr)) = row.get_value(col_name) {
            // Convert array of objects to Vec<Row>
            return arr
                .iter()
                .filter_map(|v| {
                    if let Value::Object(obj) = v {
                        // Convert HashMap<String, Value> to Row
                        Some(Row::from_map(obj.clone()))
                    } else {
                        None
                    }
                })
                .collect();
        }
    }

    // No array column found, return cloned original result
    result.iter().map(|r| (*r).clone()).collect()
}

/// Extract node_id from a result row.
pub(crate) fn extract_node_id(row: &Row) -> Option<String> {
    row.get_value("node_id").and_then(|v| match v {
        Value::Integer(i) => Some(i.to_string()),
        Value::String(s) => Some(s.clone()),
        _ => None,
    })
}

/// Extract user_id from a result row.
pub(crate) fn extract_user_id(row: &Row) -> Option<String> {
    row.get_value("user_id").and_then(|v| match v {
        Value::String(s) => Some(s.clone()),
        Value::Integer(i) => Some(i.to_string()),
        _ => None,
    })
}

/// Extract a float score from a result row.
pub(crate) fn extract_float(row: &Row, field: &str) -> f64 {
    row.get_value(field)
        .map(|v| match v {
            Value::Float(f) => *f,
            Value::Integer(i) => *i as f64,
            _ => 0.0,
        })
        .unwrap_or(0.0)
}

/// Extract an integer value from a result row.
pub(crate) fn extract_int(row: &Row, field: &str) -> i64 {
    row.get_value(field)
        .map(|v| match v {
            Value::Integer(i) => *i,
            Value::Float(f) => *f as i64,
            _ => 0,
        })
        .unwrap_or(0)
}

/// Extract a string value from a result row.
pub(crate) fn extract_string(row: &Row, field: &str) -> Option<String> {
    row.get_value(field).and_then(|v| match v {
        Value::String(s) => Some(s.clone()),
        Value::Integer(i) => Some(i.to_string()),
        _ => None,
    })
}
